Omv initial integrate

nxf48054 opened the pull request:
fix uart bug

RockySong committed:
first time add minial openMV files

RockySong committed:
first time merge minimal openMV functionality to i.MX RT1050 mpy port.

RockySong committed:
preliminary make successful communication with openMV IDE, however it does not respond to openMV IDE when executing script, need to fix.
fix pendSV handler bug: in KEIL, should not use C handler to call asm handler
merge openMV's main.c to project and created main_omv.c for necessary init and nlr_jump usage.

RockySong committed:
fix ov7725 register init issue, the "yellow effect" is removed.
Disable print to USB VCOM when openMV IDE is communicatiing with board.

RockySong committed:
fix virtual comm to implement "fragmentation" of jpeg data transfer to PC
enable jpeg software compression.
can show jpeg to openMV IDE now, but color space is wrong, and very slow. (openMV original has 23fps at qqvga, while currently this one has only 3-4 fps @qqvga)

RockySong committed:
update mpy-cross: openMV modifed nlr_raise() and add a function, which is not avaiable for mpy-cross build. So modified mpy-cross Makefile to add a compiler defined macro, and modified nlr.h to recognize the macro to disable openMV specific thing.
update ov7725 COM3 init value to make jpeg compress successful.

RockySong committed:
add image module
imlib_find_blobs() may create zero pointer that corrupts address 0x04, which is firmware code segment, lead to hard fault (6+ hours to find this bug!), not yet solved.
add floating point functions for KEIL __CC_ARM
use pseduo rand to implement a weak rng_randint() that is required by kmeans.c
Restore py_sensor_snapshot() to have the image object
Adjust struct framebuffer to make pixels[] to align to 8 bytes, which is required by i.MX RT1050's CSI.
Expand non-cachable SDRAM from 2MB to 4MB, still at the tail.
JPEG.c moved to ITCM, yuv_tab moved to non-cachable SDRAM

RockySong committed:
board.c: Rewrite MPU regions and add one region to disallow write access to bottom 8kB of ITCM, in case wild pointers corrupt code in ITCM.
isr.c: Add MemManage fault handler
omv/img/collections.c has bug to write to null pointer, fix it
sensor.c, py_sensor.c: replace to newer version, and get rid of line filter features but keep the sensor_snapshot() compatible with original OpenMV.
py_helper.c: add required functions from other version : py_helper_keyword_xxx.

RockySong committed:
update gitignore file

RockySong committed:
main.c, main_omv.c : fix bug to enable auto run main.py
usbdbg.c, usbapp.c, virtual_com.c: support openMV output only terminal during IDE is connected
ring_buffer.c: fix bugs, still not all fixed yet
put lab and xyz tab to DTCM, performance is 3x than SDRAM.

RockySong committed:
ring_buffer.c: fixed a bug of ring block, which in corner case that ring block is full, will wrongly adjust wNdx (in this case, both rNdx and wNdx is the same, and blkWNdx is 0), that leads ring block to stuck.

RockySong committed:
Copy vector table to ITCM when excuting from FlexSPI
Add FlexSPI XIP project configuration, extract flash programming code to ITCM (still under debugging)
Add "MEM_PROFILING" macro to only link minial code for easy profiling
update KEIL startup file to support flexSPI XIP debugging (to be removed)
Greately shrunk USB_DEVICE_MSC_READ_BUFF_SIZE and USB_DEVICE_MSC_WRITE_BUFF_SIZE (better to keep them multiples of 512 to ensure SD card write performance)
add an example of openMV IDE protocol.
restore automatic JPEG compression quality adjust.

RockySong committed:
added raw framebuffer transfer mode, however time save of skipping JPEG can't compensate time added for USB transfer (host always send max 64 bytes?)
bug to fix: when using raw FB transfer mode, sometimes hard fault occurs.
updated SDMMC driver to 2.3.1, better recognition to SD cards.

RockySong committed:
fb_alloc.c: fix keil scatter file bug that makes the linker ignore the OMV_FB_ALLOC Execution Region
ring_buffer.c: fixed a bug when ring blocks are just read one time after previous full, should not allow to write in the same block, though not full.
add sdram_sdr_fb.sct which put framebuffer and its extra bytes to sdram.
jpeg.c: Put all const data to ".jpeg_tab" section.
fmath: inline assembly some of fast floating math functions. In ARMCC, inline in thumb under thumb has been supported from 5.01 version.

RockySong committed:
fsl_csi.c, sensor.c: Modify to support 1 shot snapshot for single FB.
ring_buffer.c: there are bugs for long time. Fix typo bug of  "if (--maxBlks)"; Allow a partial read and written block to be continuned to write in; change to when a partial written block is exhausted (the whole RB is empty in this case), reuse this block.
add debug code for ring_buffer for openMV tx.

RockySong committed:
ring_buffer.c: fixed another bug of ring block (hope to be the last one): Disallow to write to a block from beginning if it is not empty (being partially read), otherwise "usedCnt" indicator will be wronglly calcuated.

RockySong committed:
ring_buffer.c: bugfix to correctly determine when RB is just read from full state.

RockySong committed:
a buggy temp version

RockySong committed:
add missing overlay manager
recover from hardfault caused by INVSTATE: clear the wrong "ICI/IT" field in EPSR.
still hardfault remaining.
fixed ide exception being GC'ed bug: add to root pointer.

RockySong committed:
Apply openMV author's suggestions to split compile and execute, and disable USB IRQ during compile, to avoid pendSV during mp_parse(), which triggers INVSTATE hardfault
Modify sensor driver to generate IRQ per bulk, usually one bulk is one line, unless line size is too small that can't fit to smallest CSI DMA transfer unit (32 bytes)

RockySong committed:
Changed CSI driver to just receive one bulk (usually one line, can be multi-lines for smaller resolutions) per DMA  transfer, so grayscale image capture no longer need 2 times of framebuffer size (CSI does not support auto skip of U and V, have to receive all).

RockySong committed:
micro tune

RockySong committed:
optimize Y extraction in YUV format: load 32 byte every time, reschedule asm code to unleash dual-issue of Cortex-M7

RockySong committed:
add more overlays: face detection code and RO, now put to same region, assuming ICache will work.
note: overlaid code may fool the debugger to show wrong source code, as overlaid code can have the same address.
fix typo bugs in omv_boardconfig.h
add EVK LCD preview. However, when LCD is working, sometimes debug session is interrupted, and code occasionally corrupt
virtual com: detect VCOM disconnect, on disconnect, disable JPEG FB to save time.

RockySong committed:
add flexspi operations to overlay linkage
fix xip scatter file bug about IVT placement: Must not use loadAddr!=ExeAddr for RO data, have to use dedicated load region for ivt.

RockySong committed:
add profiling in isr.c: systick calculates where is interrupted
tune scatter file to optimize performance.
fix overlay manager bug
enable SWO pin for trace.

RockySong committed:
A n intermediate commit, do not use
updated makefile
moved overlay manager source file location
added GCC version of some assembly functions.

RockySong committed:
A  intermediate buggy commit
Build image under GCC, but with many hardfaults, and SD card init is easy to fail.

RockySong committed:
Many bug fixes for GCC, mainly related to GCC specific extensions (such as naked function) and improper .ld file (makes varaiables unexpected overlaid).
Still buggy! Currently, by laying mpy heap far away from normal data and bss can reduce hardfault, there must be hidden improper .ld usage/gcc extension usages.
bug fix for GCC: Asm modules added "__attribute__((naked))", this is mandatory to avoid GCC to generate code that has side effects for ASM functions
bug fix for LD file: if .ld file misses placement of sections, LD's default placement may lead to confliction with other data, causes very hard to debug hardfaults that have different behaviors. The example is NonCacheable section. When rewriting .ld file, discarded it due to unaware of this error prone LD nature.
bug fix for PendSV handler: pendSV's final handler should work on the original pendSV stack, without indirect invoke, but code called it from a C PendSV_Handler. In higher optimization level, GCC eliminates this indirect invoke, but in lower optimization level (O1, O0), GCC leaves this indirect invoke that corrupts LR (R14), so bug only occurs in lower optimization levels!
(may fix bug): Increase USB_DEVICE_MSC_WRITE_BUFF_SIZE and USB_DEVICE_MSC_READ_BUFF_SIZE both to 2kB, this may help SD card to be recognized in a higher rate.
reduce RAM waste: layout "qh_buffer" who wants 2kB alignment to a dedicated section and put it adjacent to a 2kB aligned section, this can avoid filling up to 2kB-1 bytes of RAM ---- wasted in vain.
before fulling understanding .ld file, cancel section placement for jpeg tab and lab_tab
nlrthumb: rewrite to make GCC and __CC_ARM to share the same C function.

RockySong committed:
bug fix for GCC: GCC may not push R0-R3,LR when calling naked functions. so push them in naked functions.

RockySong committed:
virtual_com: Increase VCOM endpoint buffer from 64Byte to 128Byte
main_omv.c: Restore the split openMV script runner: compile with USB IRQ disabled, then enable USB IRQ, then execute.
Disable MICROPY_ENABLE_SCHEDULER in mpconfigport.h
(buggy?) overlay_manager.c: Rearrange code. There is a unreasonable issue: When overlay is used, script may stuck 0.5-1.5 seconds during execution. If trying to capture PC in debug session for the stuck, then under KEIL, it will say could not stop Cortex-M core and quit the debug session.
pendsv.c: "inline" the pendsv_kbd_intr()
(improve) implement fast_roundf(), fast_ceilf(), fast_floorf() for KEIL using embedded assembly and hard FP, can get great performance boost, though still not as perfect as GCC which can inline functions which has inline assembly and written in .C files.
(pending issues): Memory corruption between RW and MPY heap (lead to Precise bus fault); INVSTATE hardfault with xpSR.[13:12] = 3

RockySong committed:
restore overlay and lcd monitoring

RockySong committed:
standarize irq priorities management
In pendSV handler, modify stacked xpsr to clean state in case it interrupts LDM/STM to lead to INVSTATE hardfault
In nlr_jump_asm, modify xpsr to clean state
improve image capturing to use as much line buffer as possible, this can decrease CSI IRQ rate.
allow set_framerate() to accept raw CCM register value to set CSI clock source and pre divider, this allows QVGA snapshot at 100+ fps.
fix bugs for windowing snapshot.
(apriltag.c and dmtx.c) fix bugs that prevent from using fast_xxx floating point routines.
ring_buffer.c: critical section use BASEPRI, does not mask CSI IRQ.
sdram.sct: try to no longer put ARM floating point lib functions in ITCM.

RockySong committed:
microtune, remove DSB() at the end of USB IRQ, as it can stall M7 up to 100+ cycles.

RockySong committed:
add mjpeg module
upgrade profiling module

RockySong committed:
fix bug in profiling module and micro tune
put stack into itcm, save a bit of dtcm space, performance change is very little
allow script to control usb irq enable/disable during snapshot (s_isEnUsbIrqForSnapshot)

RockySong committed:
sensor.c: remove unused logic and code clean

RockySong committed:
partially update files

RockySong committed:
update more (test branch merging)

RockySong committed:
tiny code clean

RockySong committed:
Vividsun: test

Vividsun committed:
vivid sun test 2

Vividsun committed:
vividsun test 3

Vividsun committed:
vivid sun test 4

RockySong committed:
update more files
usbdbg: restore image save

RockySong committed:
Updated to new openMV version, face detection and blob detection works normally.
latest openMV removed some filters: mean.c, median.c, midpoint.c, mode.c, and morph.c
disable arm_const_structs.c: it depends on very large tables.
added CMSIS DSP and NN files.

RockySong committed:
change sd mount point to root dir '/'
updated to latest openMV and add machine learning features. can load cifar10_fast (~33kB model+weights) to ~48kB mpy heap

RockySong committed:
Merge branch 'omv_nn_up' into omv_initial_integrate

RockySong committed:
add missing files for ml and nn.

RockySong committed:
fix keil projects to use O3 Otime for CMSIS-NN library to get best optimizations.
Add readme.md for i.MX RT1050 openMV port.

RockySong committed:
add frozen_mpy.c, so user nolonger need to first run make then build with keil.

RockySong committed:
added RT1060 support.
sensor driver works with DCache

RockySong committed:
temp buggy, do not use : on IRQ, CM7 core lose control, debug session stops, all registers and memory values are wrong, IRQ = 1 (Reset)
add pinmux settings for sensor PWDN and RSTB signal
remove .bin file

RockySong committed:
intermediate test push, buggy, do not use!
fix ram_end address
disable uart temporaryly
disable NULL pointer capture temporily.

RockySong committed:
openMV works on omvrt1 board now, USB + SDCARD may have some issues of h/w or s/w, or both.
update scatter file to make use of extra 512kB of OCRAM on i.MX RT1060
small bug fixes

RockySong committed:
change folder name

RockySong committed:
change names, rt105,rt106 to rt1050, rt1060

RockySong committed:
delete old project folder

RockySong committed:
__WFI() in RT1060 causes debug session broken and speed change, perhaps PMU is not correctly configured, currently, change __WFI() to HAL_WFI() which is empty for workaround
fsl_sd.c always switch to 1.8V if card supports 1.8V, however omvrt1 board only supplies 3.3V, this makes various SD card init issues, fixed this issue.
Modifed SD card pin init to support stronger drive.
openMV may access overflow mpy heap, so reserve 16kB at the end of RAM (main.c)
Update omv_boardconfig.h and scatter file to make use of extra 512kB OCRAM on i.MX RT1060

RockySong committed:
bufgix: ov7725 input clock must not exceed 48MHz. Previous code violates this rule. Fixed to 15MHz input MCLK, and use OV7725's internal PLL to get 60MHz. Otherwise, We can only get 48MHz.
bugfix: Disable IRQ in _Start_USB_VCOM_Write(), otherwise the 1st byte of tx buffer may be sent twice
bugfix: Add API of "RingBlk_ReuseTakenBlk" in ring_buffer.c, and for control chars, such as CTRL-C, abandon RX data and use this API to reuse the taken block of USB OUT EP.
bugfix: clear [31:24] of int data in USB VCOM RX, for char input.
bugfix: Change __WFI() to HAL_WFI() in remaining code snippets, including MICROPY_EVEN_POOL_HOOK macro, which check exceptions and initiates nlr_raise()
Initially added ADC module.

RockySong committed:
bugfix: IR led's logic is inverted on omvrt1 board

RockySong committed:
improve vcom open/close detect: I found all my VCOM applications always send SetControlLineState on VCOM close. SetLineCoding, GetLineCoding, RX event always set VCOM to Open; SetControlLineState always sets VCOM to Close, if this is wrong, later set VCOM to open events can correct it.

RockySong committed:
Add a new keil project prj_keil_pyb_rt1050, which uses OpenMV RT code base, but exclude openMV sensor and image dependencies, intend to debug micropython and pyb related code.
Add a new project macro 'OMV_MPY_ONLY', if true, then exclude openMV sensor and image related dependencies.
Bugfix: make-pins.py has a typo that mess ADC number.

RockySong committed:
update openmv rt arch string and ID, and add note.

RockySong committed:
regulate projects:
prj_keil_rt1050: OpenMV RT board w/ i.MX RT1052 (not ready)
prj_keil_rt1060: OpenMV RT board w/ i.MX RT1062
prj_keil_rt1050_evk: EVK w/ i.MX RT1052
prj_keil_rt1060_evk: EVK w/ i.MX RT1062 (not ready)
prj_keil_pyb_rt1050: OpenMV RT board w/ either 1052/62, but removed openmv features except usbdbg with openmv ide, used for internal mpy development and testing

nxf48054 committed:
fix the uart bug and we can use it transmit the data

nxf48054 committed:
refresh the project and add some missing file

RockySong committed:
add 'rocky_test' folder, the contents of it can be put to a TF card for demo color blob, face detect, cifar-10, lenet.
Fix project files

RockySong reviewed (commented):
Already merged into omv_initial_integrate, not 'master'.

RockySong committed:
shrink framebuffer and fb_alloc from 365K/144K to 301K/83K, which is the same size as OpenMV3. Leave some margin for other future features.
Added SPI class for omvrt1 board.

RockySong committed:
Add KEIL flash loader for GD25Q32

RockySong committed:
add i2c class.
update uart.h to adapt to different boards.
remove .uvgui.xx files from repo.

RockySong committed:
update nn_convert.py to support python 3
added nn_convert_txt.py to generate a human readable network file.

RockySong committed:
sensor.c: Some RW / ZI cause cache related bug: if script uses VGA  by default then hardfault. Move sensor.c RW and ZI to DTCM.
moved usbd_msc_storage.c RW and ZI to OCRAM

RockySong committed:
improve multiple demos
fix rt1050 evk project: use SDRAM to work as framebuffer and jpeg buffer.
fix rt1050 evk lcd review: flip horizontal to make OSD correct, but will mirror image

RockySong committed:
Port to RT1060 EVK A1 board.
Delete "prj_keil_rt1060_evk", and make "prj_keil_rt1050_evk" can support both 1050 and 1060.
add SDRAM framebuffer support for rt1050/60 EVK, add IS25WP064 (QSPI flash) support for rt1050/60 EVK.
improve demos: add demos that uses SDRAM ---- larger frame size

RockySong committed:
multi_demos_sdram.py: qrcode demo does not work after other demos run, so add machine.reset() every round.

RockySong committed:
add RW and ZI of virtual_com.c to DTCM to avoid cache inconsistency, which causes openmv ide connection errors.
micro tune multiple-in-1 demo scripts of openmv

RockySong committed:
bugfix: Apriltag recognition failure: Both CPU and DMA write to the same cache line (Header and data of MAIN_FB()), but "DCACHE_InvalidateByRange()" is wrongly used instead of "DMA_CleanInvalidateByRange()". Fixed it by use DMA_CleanInvalidateByRange() in CSI IRQ handler; use DMA_CleanByRange() when start a new frame, and disable bufferable ---- write allocate in MPU configuration of OCRAM. In fact any one of the 3 can fix this issue.

RockySong committed:
add cache erratum test code and enable bufferable to OCRAM to avoid the erratum.

RockySong committed:
Make the cache erratum test more graceful

RockySong committed:
SD card: Never switch to 1.8V
flashpgm: only init flashfs and overlay flashpgm code for hyper flash on evk
cmos sensor: use lower sensor clock for EVK
mpu setup: Allow write to first 8KB from 0; 3.5MB-4MB offset of flexspi disables cache

RockySong committed:
nn.c: fixed a bug that missed to count col_buf size for IP/FC layer.
usb sd disk: Increase USB_DEVICE_MSC_READ_BUFF_SIZE from 2kB to 4kB

RockySong committed:
nn.c: Found previous ip layer colbuf bug is fixed, use openmv official fix

RockySong committed:
added rtc, rng, unique id drivers, thanks Curry Liu!
mpu config: For imxrtevk board: offset 3.5MB-4MB of FlexSPI is changed to non-cachable.

RockySong committed:
tiny tune
bug fix: gpio output flag overflows mpy's int
virtual_com: "WriteAlways" no longer polling for VCOM

RockySong committed:
forgot

RockySong committed:
minor code clean
remove the test python module "tpyboard"

RockySong committed:
fix bugs for pin and spi
fsl_lpspi: updated to poll LPSPI's busy flag to work with manual CS control (as GPIO)
mutext.c: Give OpenMV IDE more chance to lock jpeg buffer

RockySong committed:
bugfix for OpenMV IDE connection slow or stuck
fix rng methohd
add gif module

RockySong committed:
fix the schematics pdf file omvrt1_sch.pdf
Change board pin naming so it is compliant with OpenMV Cam M7/H7
add '信号定义.xlsx' for expension pin function definitions.

RockySong committed:
Generate bits/bitfields definitions within peripheral registers as QSTRs (make-mcuconst.py added '-bf' argument, Makefile)

CristXu committed:
add some self file

CristXu committed:
update the modpyb.c

RockySong committed:
Merge pull request #10 from nxf48054/omv_initial_integrate

Added preliminary PWM class to support PWM1-3 from Crist Xu.

RockySong committed:
Add pwm.c to keil project, add AF_PIN_TYPE_PWM_PWMxx enums to pin_defs_mcu.h

RockySong committed:
Update README.md

RockySong committed:
make-mcuconst.py can generate bits and bitfields inside registers.

RockySong committed:
Update README.md

RockySong committed:
openmv version upgrade

RockySong committed:
upgraded sensor.c (partially, mainly for ioctl, palette), framebuffer (struct typedef)
enlarge fb_alooc buffer from 84KB to 120KB
upgraded apriltag

RockySong committed:
upgraded imlib, py_image, blob, binary, rainbow
fmath.h: uses inline assembly in CC ARM, hopefully faster.
added selective_search.c, wifidbg.c (not yet ported), py_tv.c (not yet used)
removed lenet.c, lenet_model_num.c

RockySong committed:
bugfix: if 'main.py' is running when executing another script from OpenMV IDE, fb alloc buffer may fail, fixed by soft reset when staring a new script in OpenMV IDE, the cost is user have to press start button twice to start it.
upgrade all upgradable OpenMV firmware to 3.3.1

RockySong committed:
add more comments

RockySong committed:
Update README.md

Added lots of introduction in Chinese.

RockySong committed:
added rpm and pwm drivers, still preliminary.

RockySong committed:
mphalport.c: fix gpio config bug: manually enable input path of GPIO, by setting bit 4 (SION) of corresoponding MUX register of GPIO's pin ; also set this bit for open-drain modes
mphalport.h: fix gpio pad read bug: read "PSR" register instead of read "DR" register. As DR just reflects the data to drive, but not the level of pad, though in most (normal) cases the two are equivalent.
pin.c: correct LED pin orders
pin_defs_mcu.h: Fix a bug for open drain mode: Must set pad's output strength to non-zero
shrink jpeg buffer size from 32KB to 24KB
i2c.c: fix scan() bug by adding a safe delay after sending out the start conditon and device address. LPI2C_CheckForBusyBus() is not working, though it seems should works, cost several hours! damn it!
update config for RT1052: shrink fb size to 200KB, enlarge min fb_alloc size to 120KB, cancel overlay for YUV and LAB tab. This has minor slow down of JPEG compress, significant (>=240%) slow down for color blob searching.
multi_demos_max.py: shrink frame size for QRcode detect demo so RT1052 can run it.

RockySong committed:
workaround I2C NAK issue by using volatile loop delay
fix pin print bug and print board name with better format

RockySong committed:
Have to add more time for I2Cscan safe delay
better pin print: print pad level if SION bit (bit4) in MUX CFG reg is set.

RockySong committed:
add qtimer class from Crist Xu, (experimental), implements "timer" class from pyb module, except the "channel" support which is bounded to STM32. Use 11 qtimer sub modules to provide short period, high precision timing.
support 4 rpms
qtimer: automaticall stop all callbacks when script is stopped
enable mpy scheduler for callback's need.

RockySong committed:
merge pwm.c from Crist to add more PWM channels, and preliminary servo
add preliminary DC motor control (dcmc.c, dcmc.h)

RockySong committed:
Update README.md

correct I2C number, I2C4 -> I2C3, thanks Ericir3

RockySong committed:
fix a CSI driver bug, thanks Dean Jia!

RockySong committed:
Update multi_demos_max.py

correct typo

RockySong committed:
Update README.md

RockySong committed:
Update README.md

RockySong committed:
Cancel overlay for jpeg compression code and YUV tab
bug fix: Some LAB tab overlay load is missing, one led to color image binary() failed, added them.

RockySong committed:
added experimental Servo, srpm, tmr (software timer) module.
rewrite QTimer IRQ handler to support both "QTIMER" (high resolution but short max period h/w timer) class and "Servo" class.
Updated pyb only keil project.

RockySong committed:
Restored gnerated files for evk board

RockySong committed:
Update README.md

Add 2 FAQs, 
Fix a critical typo "N<<12" to "N<<11"
